File: tools/test_db_download.swift
#!/usr/bin/env swift

import Foundation

// GitHub 설정
let githubRepo = "SmallTyrant/hololive_OCG_helper"
let dbReleaseTag = "DB"
let dbReleaseAPI = URL(string: "https://api.github.com/repos/\(githubRepo)/releases/tags/\(dbReleaseTag)")!
let dbDirectURL = URL(string: "https://github.com/\(githubRepo)/releases/download/\(dbReleaseTag)/hololive_ocg.sqlite")!

struct ReleaseDbInfo {
    let tag: String
    let assetName: String
    let assetURL: URL
    let assetUpdatedAt: String
    let publishedAt: String
    let createdAt: String
}

// 테스트 1: GitHub API 호출
func testGitHubAPI() async throws {
    print("📡 테스트 1: GitHub API 호출")
    print("   URL: \(dbReleaseAPI.absoluteString)")
    
    var request = URLRequest(url: dbReleaseAPI)
    request.timeoutInterval = 20
    request.setValue("hOCG_H/1.1", forHTTPHeaderField: "User-Agent")
    request.setValue("application/vnd.github+json", forHTTPHeaderField: "Accept")
    
    let (data, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else {
        throw URLError(.badServerResponse)
    }
    
    print("   ✅ HTTP 상태 코드: \(http.statusCode)")
    
    guard (200..<300).contains(http.statusCode) else {
        throw URLError(.badServerResponse)
    }
    
    let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any]
    guard let payload else {
        throw URLError(.cannotParseResponse)
    }
    
    let tag = payload["tag_name"] as? String ?? "unknown"
    let assets = payload["assets"] as? [[String: Any]] ?? []
    
    print("   ✅ 태그: \(tag)")
    print("   ✅ Assets 개수: \(assets.count)")
    
    // Asset 찾기
    var foundAsset: (name: String, url: String)?
    for item in assets {
        let name = item["name"] as? String ?? ""
        let urlString = item["browser_download_url"] as? String ?? ""
        if name == "hololive_ocg.sqlite" {
            foundAsset = (name, urlString)
            print("   ✅ DB Asset 발견: \(name)")
            print("   ✅ 다운로드 URL: \(urlString)")
            break
        }
    }
    
    guard foundAsset != nil else {
        throw URLError(.fileDoesNotExist)
    }
}

// 테스트 2: DB 파일 다운로드
func testDBDownload() async throws {
    print("\n📥 테스트 2: DB 파일 다운로드")
    print("   URL: \(dbDirectURL.absoluteString)")
    
    var request = URLRequest(url: dbDirectURL)
    request.timeoutInterval = 120
    request.setValue("hOCG_H/1.1", forHTTPHeaderField: "User-Agent")
    request.setValue("application/octet-stream", forHTTPHeaderField: "Accept")
    
    let (tempURL, response) = try await URLSession.shared.download(for: request)
    guard let http = response as? HTTPURLResponse else {
        throw URLError(.badServerResponse)
    }
    
    print("   ✅ HTTP 상태 코드: \(http.statusCode)")
    
    guard (200..<300).contains(http.statusCode) else {
        throw URLError(.badServerResponse)
    }
    
    let fm = FileManager.default
    let attrs = try fm.attributesOfItem(atPath: tempURL.path)
    let fileSize = attrs[.size] as? Int64 ?? 0
    
    print("   ✅ 다운로드 완료: \(fileSize) bytes (\(Double(fileSize) / 1024.0 / 1024.0) MB)")
    
    // SQLite 헤더 검증
    let data = try Data(contentsOf: tempURL, options: .mappedIfSafe)
    guard data.count > 16 else {
        throw URLError(.cannotDecodeContentData)
    }
    
    let header = Data("SQLite format 3\u{0}".utf8)
    guard data.prefix(16) == header else {
        throw URLError(.cannotDecodeContentData)
    }
    
    print("   ✅ SQLite 헤더 검증 성공")
}

// 테스트 3: Fallback URL 검증
func testFallbackURL() async throws {
    print("\n🔄 테스트 3: Fallback URL 검증")
    print("   Fallback URL: \(dbDirectURL.absoluteString)")
    
    var request = URLRequest(url: dbDirectURL)
    request.httpMethod = "HEAD"
    request.timeoutInterval = 20
    request.setValue("hOCG_H/1.1", forHTTPHeaderField: "User-Agent")
    
    let (_, response) = try await URLSession.shared.data(for: request)
    guard let http = response as? HTTPURLResponse else {
        throw URLError(.badServerResponse)
    }
    
    print("   ✅ HTTP 상태 코드: \(http.statusCode)")
    
    guard (200..<300).contains(http.statusCode) else {
        throw URLError(.badServerResponse)
    }
    
    print("   ✅ Fallback URL 접근 가능")
}

// 메인 실행
Task {
    print("🧪 DB 다운로드 기능 테스트 시작\n")
    print(String(repeating: "=", count: 60))
    
    do {
        try await testGitHubAPI()
        try await testDBDownload()
        try await testFallbackURL()
        
        print("\n" + String(repeating: "=", count: 60))
        print("✅ 모든 테스트 통과!")
        exit(0)
    } catch {
        print("\n" + String(repeating: "=", count: 60))
        print("❌ 테스트 실패: \(error)")
        if let urlError = error as? URLError {
            print("   URLError 코드: \(urlError.code.rawValue)")
            print("   설명: \(urlError.localizedDescription)")
        }
        exit(1)
    }
}

dispatchMain()

File: tools/test_db_download.py
#!/usr/bin/env python3
"""
Android UpdateRepository 로직을 Python으로 재현하여 테스트
"""

import json
import hashlib
import sys
import urllib.request
from urllib.error import HTTPError, URLError

# GitHub 설정
GITHUB_REPO = "SmallTyrant/hololive_OCG_helper"
DB_RELEASE_TAG = "DB"
DB_RELEASE_API = f"https://api.github.com/repos/{GITHUB_REPO}/releases/tags/{DB_RELEASE_TAG}"
DB_DIRECT_URL = f"https://github.com/{GITHUB_REPO}/releases/download/{DB_RELEASE_TAG}/hololive_ocg.sqlite"

def normalize_hash(raw):
    value = (raw or "").strip().lower()
    if value.startswith("sha256:"):
        value = value[len("sha256:"):].strip()
    return value

def get_release_asset():
    request = urllib.request.Request(
        DB_RELEASE_API,
        headers={
            "User-Agent": "hOCG_H/1.1",
            "Accept": "application/vnd.github+json"
        }
    )

    with urllib.request.urlopen(request, timeout=20) as response:
        if response.status < 200 or response.status >= 300:
            raise Exception(f"HTTP 오류: {response.status}")
        data = json.loads(response.read().decode("utf-8"))

    assets = data.get("assets", [])
    for asset in assets:
        if asset.get("name") == "hololive_ocg.sqlite":
            return {
                "tag": data.get("tag_name", "unknown"),
                "name": asset.get("name", ""),
                "url": asset.get("browser_download_url", DB_DIRECT_URL),
                "digest": normalize_hash(asset.get("digest", "")),
            }

    raise Exception("DB asset을 찾을 수 없습니다")

def test_github_api():
    """테스트 1: GitHub API 호출"""
    print("📡 테스트 1: GitHub API 호출")
    print(f"   URL: {DB_RELEASE_API}")
    
    request = urllib.request.Request(
        DB_RELEASE_API,
        headers={
            "User-Agent": "hOCG_H/1.1",
            "Accept": "application/vnd.github+json"
        }
    )
    
    try:
        with urllib.request.urlopen(request, timeout=20) as response:
            print(f"   ✅ HTTP 상태 코드: {response.status}")
            
            data = json.loads(response.read().decode('utf-8'))
            tag = data.get("tag_name", "unknown")
            assets = data.get("assets", [])

            print(f"   ✅ 태그: {tag}")
            print(f"   ✅ Assets 개수: {len(assets)}")

            asset = get_release_asset()
            print(f"   ✅ DB Asset 발견: {asset['name']}")
            print(f"   ✅ 다운로드 URL: {asset['url']}")
            print(f"   ✅ Digest: {asset['digest'] or '(없음)'}")
                
    except HTTPError as e:
        raise Exception(f"HTTP 오류: {e.code} - {e.reason}")
    except URLError as e:
        raise Exception(f"URL 오류: {e.reason}")

def test_db_download():
    """테스트 2: DB 파일 다운로드"""
    asset = get_release_asset()
    print("\n📥 테스트 2: DB 파일 다운로드")
    print(f"   URL: {asset['url']}")
    
    request = urllib.request.Request(
        asset["url"],
        headers={
            "User-Agent": "hOCG_H/1.1",
            "Accept": "application/octet-stream"
        }
    )
    
    try:
        with urllib.request.urlopen(request, timeout=120) as response:
            print(f"   ✅ HTTP 상태 코드: {response.status}")
            
            data = response.read()
            file_size = len(data)
            print(f"   ✅ 다운로드 완료: {file_size} bytes ({file_size / 1024 / 1024:.2f} MB)")
            
            # SQLite 헤더 검증
            if len(data) < 16:
                raise Exception("파일이 너무 작습니다")
            
            header = b"SQLite format 3\x00"
            if data[:16] != header:
                raise Exception("유효한 SQLite 파일이 아닙니다")
            
            print("   ✅ SQLite 헤더 검증 성공")

            actual_digest = hashlib.sha256(data).hexdigest()
            expected_digest = asset["digest"]
            if expected_digest:
                if actual_digest != expected_digest:
                    raise Exception("Digest 검증 실패")
                print(f"   ✅ SHA256 검증 성공: {actual_digest}")
            else:
                print("   ℹ️  Digest 정보 없음 - SHA256 검증 생략")
            
    except HTTPError as e:
        raise Exception(f"HTTP 오류: {e.code} - {e.reason}")
    except URLError as e:
        raise Exception(f"URL 오류: {e.reason}")

def test_fallback_url():
    """테스트 3: Fallback URL 검증"""
    print("\n🔄 테스트 3: Fallback URL 검증")
    print(f"   Fallback URL: {DB_DIRECT_URL}")
    
    request = urllib.request.Request(
        DB_DIRECT_URL,
        headers={"User-Agent": "hOCG_H/1.1"},
        method="HEAD"
    )
    
    try:
        with urllib.request.urlopen(request, timeout=20) as response:
            print(f"   ✅ HTTP 상태 코드: {response.status}")
            print("   ✅ Fallback URL 접근 가능")
            
    except HTTPError as e:
        raise Exception(f"HTTP 오류: {e.code} - {e.reason}")
    except URLError as e:
        raise Exception(f"URL 오류: {e.reason}")

def main():
    print("🧪 DB 다운로드 기능 테스트 시작 (Android 로직)\n")
    print("=" * 60)
    
    try:
        test_github_api()
        test_db_download()
        test_fallback_url()
        
        print("\n" + "=" * 60)
        print("✅ 모든 테스트 통과!")
        return 0
    except Exception as e:
        print("\n" + "=" * 60)
        print(f"❌ 테스트 실패: {e}")
        return 1

if __name__ == "__main__":
    sys.exit(main())
